### PR TITLE
Issue #1342 Moving the default add-on install to start

### LINKED ISSUE #1342: Default add-ons install message should not come for all commands
```
$ make integration GODOG_OPTS="-tags command -format pretty"
go install -pkgdir=/home/amit/go/src/github.com/minishift/minishift/out/bindata -ldflags="-X github.com/minishift/minishift/pkg/version.minishiftVersion=1.5.0 -X github.com/minishift/minishift/pkg/version.b2dIsoVersion=v1.1.0 -X github.com/minishift/minishift/pkg/version.centOsIsoVersion=v1.1.0 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v3.6.0 -X github.com/minishift/minishift/pkg/version.commitSha=0e75c4ec" ./cmd/minishift
mkdir -p /home/amit/go/src/github.com/minishift/minishift/out/integration-test
go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration -v -args --test-dir /home/amit/go/src/github.com/minishift/minishift/out/integration-test --binary /home/amit/go/bin/minishift -tags command -format pretty
Test run using Boot2Docker iso image.
Keeping Minishift cache directory '/home/amit/go/src/github.com/minishift/minishift/out/integration-test/cache' for test run.
Log successfully started, logging into: /home/amit/go/src/github.com/minishift/minishift/out/integration-test/integration.log
Running Integration test in: /home/amit/go/src/github.com/minishift/minishift/out/integration-test
Using binary: /home/amit/go/bin/minishift
Feature: Basic
  As a user I can perform basic operations of Minishift and OpenShift

Feature: Openshift commands
  Commands "minishift openshift [sub-command]" are used for interaction with Openshift
  cluster in VM provided by Minishift.

Feature: Version command
  Command "minishift version" shows version of minishift binary.

  Scenario: User can get information about version of Minishift                  # features/cmd-version.feature:5
    When executing "minishift version" succeeds                                  # integration_test.go:652 -> github.com/minishift/minishift/test/integration.executingMinishiftCommandSucceedsOrFails
    Then stdout should match "^minishift v[0-9]\.[0-9]\.[0-9]\+[a-z0-9]{7,8}\n$" # integration_test.go:630 -> github.com/minishift/minishift/test/integration.commandReturnShouldMatchRegex
    Output did not match. Expected: '^minishift v[0-9]\.[0-9]\.[0-9]\+[a-z0-9]{7,8}\n$', Actual: '-- Installing default add-ons ... OK
minishift v1.5.0+0e75c4ec
'

Feature: Cool Store
  In order to test Minishift under load
  I need to setup a test environment
  Using CentOS and 4G of memory
  After which I will import image streams
  Finally I will deploy the Cool Store

Feature: Flags
  As a user I can perform advanced operations
  such as starting Minishift with optional features

Feature: Provision an older major release
  As a user I can provision an older major version of minishift

Feature: Minishift can run behind proxy
  As a user I can use Minishift behind a proxy.
  INFO: Code behind this feature will start a proxy server in background
  on port 8181 and will try to get IP of host machine automatically
  to set MINISHIFT_HTTP_PROXY for this feature only.
  However one can override port number or set IP manually if needed.
  To use custom port number run this test with environmental variable
  INTEGRATION_PROXY_CUSTOM_PORT.
  To use custom IP address please set INTEGRATION_PROXY_CUSTOM_IP
  environmental variable.

--- Failed scenarios:

    features/cmd-version.feature:7

1 scenarios (1 failed)
2 steps (1 passed, 1 failed)
67.708103ms
testing: warning: no tests to run
PASS
exit status 1
FAIL    github.com/minishift/minishift/test/integration 0.085s
make: *** [Makefile:176: integration] Error 1
12:46 $
```

### PR DESCRIPTION
**Pull request by LalatenduMohanty**
Fixes #1342 

Signed-off-by: Lalatendu Mohanty <lmohanty@redhat.com>

### CODE REVIEW

File: cmd/minishift/cmd/start.go [hunk 1]
```diff
@@ -161,6 +163,12 @@ func runStart(cmd *cobra.Command, args []string) {
 	// we need to determine whether this is a restart prior to potentially creating a new VM
 	isRestart := cmdUtil.VMExists(libMachineClient, constants.MachineName)
```
Comment (gbraad) [inline cmd/minishift/cmd/start.go]: Can we move this one as much as possible to the top of the function?

This file now contains three `isRestart`, but we use them all as `!isRestart`... so maybe we should invert the meaning of the variable to something like `isFirstStart`? WDYT

Comment (LalatenduMohanty) [reply cmd/minishift/cmd/start.go]: @gbraad I think isRestart is fine and communicates the right intent.

Comment (LalatenduMohanty) [reply cmd/minishift/cmd/start.go]: On second thought I guess you meant `isFreshStart` I am fine with both isRestart and isFreshStart.

Comment (gbraad) [reply cmd/minishift/cmd/start.go]: The most important is to move it to the top of the function, as this is a conditional statement which is important for the rest of the execution.

Comment (gbraad) [reply cmd/minishift/cmd/start.go]: not a huge fan of inverted logic... especially when this is the only way we use it. inversion can lead to trouble in semantics (reading it and reasoning about it)

**Comment by LalatenduMohanty** [comment]
Need to fix the integration tests as the add-ons will not be installed before `minishift start`.

File: test/integration/features/cmd-addons.feature [hunk 2]
```diff
@@ -8,9 +8,12 @@ Feature: Addons command and its subcommands
...
 
-  @minishift-only
-  Scenario: Default add-ons are installed but not enabled by default
-  Note: default addons were installed during previous commands automatically.
+ Scenario: Installing default add-ons manually
+     When executing "minishift addons install --defaults" succeeds
+     Then stdout should contain
+      """
+      Default add-ons anyuid, admin-user, xpaas, registry-route installed
+      """
```
Comment (LalatenduMohanty) [inline test/integration/features/cmd-addons.feature]: @agajdosi @amitkrout FYI The add-on list default behavior is changing with this PR.

**Comment by hferentschik** [comment]
I need to catch up on the context and I am currently still in transit, but there is a good reason why the default add-on install is not occurring during start. There are several use cases where one want them installed prior to calling start. For example I might want to enable a given add-on or set it's priority prior to starting. There is also the interaction with the _update_ command. This has been discussed before afair. Not sure what the problem is here, but it seems wrong to me.

**Comment by LalatenduMohanty** [comment]
> There is also the interaction with the update command. This has been discussed before afair

The update context is different and the code is still present in root.go and there is not change in behavior.

**Comment by LalatenduMohanty** [comment]
> I need to catch up on the context and I am currently still in transit, but there is a good reason why the default add-on install is not occurring during start. There are several use cases where one want them installed prior to calling start. For example I might want to enable a given add-on or set it's priority prior to starting.

Sure lets discuss when you are here. But the behavior around when to execute is not changing. Also `minishift addons install --defaults` behavior does not change.

**Comment by gbraad** [comment]
An explicit command for this use case 'before start' seems more fitting than having side-effects occur on commands like `m version`, `m status`, `m ip`, `m profile`, `m config`, etc, etc.

**Comment by gbraad** [comment]
Logic should installs the addons 'before the start' of the VM, pretty much right before the preflightChecks. Nothing of libmachine is triggered at that time, so no VM exists

**Comment by gbraad** [comment]
@budhrg @praveenkumar do you remember why `root.go` was needed?
@budhrg can you check if this in any way would influence the update scenario?

@LalatenduMohanty can you repush (so the merge disappears? "Update branch" seems broken these days)

**Comment by gbraad** [comment]
@budhrg unchanged

however, `Default add-ons will be updated the next time you run any 'minishift' command.` will likely have to change going forward. WDYT? `Any` => `m addons` and `m start` would be understandable.

**Comment by gbraad** [comment]
Any update or comments? if not, I will merge today...

File: test/integration/features/cmd-addons.feature (hunk 2 above)
Comment (agajdosi) [reply test/integration/features/cmd-addons.feature]: @LalatenduMohanty Sorry I am replying late. This change is correct, in other features we use addons only in basic and there is seems that this PR won't affect the test.

**Comment by gbraad** [comment]
Merged with 349f4084b706e55e96cc536e441b161a7c6c2f87

**Comment by hferentschik** [comment]
@gbraad why did you merge? I was still having a discussion with @LalatenduMohanty? The behaviour did change for the first time install. In this case the default add-ons don't get installed in time. I suggest to reopen this issue.

**Comment by hferentschik** [comment]
> do you remember why root.go was needed?

Because you might want to interact with add-ons prior (enable, disable, change priorities) priror to starting the VM. _root.go_ is the place where we ensure the whole Minishift environment is in a consistent state. Add-ons are not tied to a 'minishift start'. If the whole issues is related to some (unwanted) output, let's talk about this and try to find a solution for that.

**Comment by gbraad** [comment]
Can you provide a testcase which indicates this situation? As at the moment, none of the errors you have mentioned have been triggered in a test or broke integration. This is undocumented behaviour.

> Add-ons are not tied to a 'minishift start'.

If this the case, they should have a place inside the `m addon` command and not inside root.  Default addons, however, need to be installed as part of the profile, prior to the actual start of the VM. This is happening...

**Comment by hferentschik** [comment]
> Can you provide a testcase which indicates this situation? As at the moment, none of the errors you have mentioned have been triggered in a test or broke integration. 

That does not mean that the behaviour has not changed. Take this (clean install, no prior Minishift)

    $ minishift add-on enable anyuid
    $ minishift start

This used to work, it does not anymore, since now the default add-ons get installed during start which is too late.

> If this the case, they should have a place inside the addon command and not inside root.

If you place it there you need to handle the default add-on install twice. Once in start and once in the addon command. PersistentPreRun is the one place where we can make sure that the Minishift install/config is complete and consistent. Hence the update functionality is there as well. 

> Default addons, however, need to be installed as part of the profile, prior to the actual start of the VM and this seems more logical to happen from the start itself. 

That's what I keep saying, it is not bound to the start command. If it were, I would not have a problem placing it there. However, there are use cases and workflows which need them installed earlier. ATM, a user on a clean system cannot configure the add-ons he wants prior to the first start, even if he wanted to.

**Comment by gbraad** [comment]
> This used to work, it does not anymore, since now the default add-ons get installed during start which is too late.

I am not following you with 'too late'. What do you expect happens?

`minishift add-on enable anyuid` only enables the install, while `minishift start` now triggers the install of the add-on still right before the VM actually starts. I am not seeing the point of it being 'too late'

> ATM, a user on a clean system cannot configure the add-ons he wants prior to the first start, even if he wanted to.

Then I would suggest to move this to `m addon` and the Prerun, but not as part of `root.go`... or at least well documented and commented.

**Comment by hferentschik** [comment]
> minishift add-on enable anyuid only enables the install, while minishift start now triggers the install of the add-on still right before the VM actually starts. I am not seeing the point of it being 'too late'

But now you cannot even enable it. `minishift add-on enable anyuid` will fail, because the default add-ons are not installed. They are now only installed once 'start' is called which is too late. 

> Then I would suggest to move this to m addon and the Prerun, but not as part of root.go

Not quite sure what you are saying. But the actual functionality to install/unpack the default add-ons is already in a shared utility package -  util.UnpackAddons. It's about where this is called. _PersistentPreRun_ is part of root.go and belongs to the root command. We discussed many times that this is the place where one needs to ensure the "correctness" of the Minishift setup. I guess we should have also added a comment, but even more that this I think we need to document the bigger architectural decisions. I propose to start using "Lightweight Architecture Decision Records" - https://www.thoughtworks.com/radar/techniques/lightweight-architecture-decision-records. Happy to get one started, but I guess this is a different issue.

**Comment by gbraad** [comment]
> because the default add-ons are not installed. 

installed => unpacked... ah, check
missing test

**Comment by gbraad** [comment]
>  I guess we should have also added a comment, but even more that this I think we need to document the bigger architectural decisions

+1

**Comment by hferentschik** [comment]
> installed => unpacked... ah, check

Right.

Also as it stands now, you still have the unwanted message in case of a Minishift _update_ command. The issue is the message not that we unpacked the add-ons in the persistent pre-run.

**Pull request closed without merging**